### PR TITLE
Allow specifying extra options for PBKDF when creating LUKS2

### PR DESCRIPTION
**Pull request by vojtechtrefny**
We need to be able to specify different kdf (e.g. pbkdf2) and
change some default parameters for argon2.

### CODE REVIEW

File: src/lib/plugin_apis/crypto.api [hunk 1]
```diff
@@ -103,9 +187,35 @@ void bd_crypto_luks_extra_free (BDCryptoLUKSExtra *extra) {
...
+ * bd_crypto_luks_extra_new: (constructor)
+ * @data_alignment: data alignment in sectors, 0 for default/auto detection
+ * @data_device: (allow-none): detached encrypted data device or NULL
+ * @integrity: (allow-none): integrity algorithm (e.g. "hmac-sha256") or NULL for no integrity support
+ * @sector_size: encryption sector size, 0 for default (512)
+ * @label: (allow-none): LUKS header label or NULL
+ * @subsystem: (allow-none): LUKS header subsystem or NULL
+ * @pbkdf: (allow-none): key derivation function specification or NULL for default
+ *
+ * Returns: (transfer full): a new LUKS xtra argument
```
Comment (vpodzime) [inline src/lib/plugin_apis/crypto.api]: *extra*

File: src/plugins/crypto.c [hunk 2]
```diff
@@ -66,6 +66,35 @@
  * Sizes are given in bytes unless stated otherwise.
  */
 
+BDCryptoLUKSPBKDF* bd_crypto_luks_pbkdf_copy (BDCryptoLUKSPBKDF *pbkdf) {
+
+    BDCryptoLUKSPBKDF *new_pbkdf = g_new0 (BDCryptoLUKSPBKDF, 1);
```
Comment (vpodzime) [inline src/plugins/crypto.c]: Extra newline here.

File: src/plugins/crypto.c [hunk 3]
```diff
@@ -462,6 +506,68 @@ gchar* bd_crypto_luks_status (const gchar *luks_device, GError **error) {
...
+    struct crypt_pbkdf_type *new_pbkdf = NULL;
+
+    if (user_pbkdf == NULL || user_pbkdf->type == NULL)
+        return NULL;
+
+    /* crypt_get_pbkdf_default returns default pbkdf parameters only based
+       on the luks version -- so for LUKS2 it returns default values for
+       argon2 but we also need to be able to provide default values if user
+       wants pbkdf2 and only specifies type -- we will use the defaults for
+       argon2 and ignore parmaters specific for it
```
Comment (vpodzime) [inline src/plugins/crypto.c]: *parameters* specific *to* it

File: src/plugins/crypto.c [hunk 4]
```diff
@@ -462,6 +506,68 @@ gchar* bd_crypto_luks_status (const gchar *luks_device, GError **error) {
...
+    else
+        new_pbkdf->iterations = default_pbkdf->iterations;
+
+    /* 'max_memory_kb' and 'parallel_threads' are not used in pbkdf2 */
+    if (g_strcmp0 (user_pbkdf->type, CRYPT_KDF_PBKDF2) == 0) {
+        if (user_pbkdf->max_memory_kb)
+            bd_utils_log (LOG_WARNING, "'max_memory_kb' is not valid option for 'pbkdf2', ignoring.");
+
+        new_pbkdf->max_memory_kb = 0;
+        new_pbkdf->parallel_threads = 0;
```
Comment (vpodzime) [inline src/plugins/crypto.c]: Why is ``parallel_threads`` not supported by ``BDCryptoLUKSPBKDF``? Is it just a matter of time somebody asks for this?

File: src/plugins/crypto.c [hunk 5]
```diff
@@ -462,6 +506,68 @@ gchar* bd_crypto_luks_status (const gchar *luks_device, GError **error) {
...
+    } else {
+        if (user_pbkdf->max_memory_kb)
```
Comment (vpodzime) [inline src/plugins/crypto.c]: So there's only PKDF2 and argon2?

File: src/plugins/crypto.h [hunk 6]
```diff
@@ -64,6 +64,27 @@ typedef enum {
...
 } BDCryptoLUKSVersion;
 
+/**
+ * BDCryptoLUKSPBKDF
+ * @type: PBKDF algorithm
+ * @hash: hash for LUKS header or NULL
+ * @max_memory_kb: requested memory cost (in KiB) or 0 for default (benchmark)
+ * @iterations: requested iterations or 0 for default (benchmark)
+ * @time_ms: requested time cost or 0 for default (bechmark)
+ * @parallel_threads: requested parallel cost (threads) or 0 for default (bechmark)
```
Comment (tbzatek) [inline src/plugins/crypto.h]: typos on the past two lines :-)

File: src/lib/plugin_apis/crypto.api [hunk 7]
```diff
@@ -103,9 +192,35 @@ void bd_crypto_luks_extra_free (BDCryptoLUKSExtra *extra) {
...
+ */
+BDCryptoLUKSExtra* bd_crypto_luks_extra_new (guint64 data_alignment, const gchar *data_device, const gchar *integrity, guint64 sector_size, const gchar *label, const gchar *subsystem, BDCryptoLUKSPBKDF *pbkdf) {
+    BDCryptoLUKSExtra *ret = g_new0 (BDCryptoLUKSExtra, 1);
+    ret->integrity = g_strdup (integrity);
+    ret->data_alignment = data_alignment;
+    ret->data_device = g_strdup (data_device);
+    ret->sector_size = sector_size;
+    ret->label = g_strdup (label);
+    ret->subsystem = g_strdup (subsystem);
+    ret->pbkdf = bd_crypto_luks_pbkdf_copy (pbkdf);
```
[the remark this answers is not in the document]
Comment (vojtechtrefny) [reply src/lib/plugin_apis/crypto.api]: Fixed in `bd_crypto_luks_pbkdf_copy`

File: src/plugins/crypto.c [hunk 8]
```diff
@@ -563,8 +685,19 @@ static gboolean luks_format (const gchar *device, const gchar *cipher, guint64 k
...
+            GError *loc_error = NULL;
             struct crypt_params_luks2 params = ZERO_INIT;
-            params.pbkdf = NULL;
+            struct crypt_pbkdf_type *pbkdf = get_pbkdf_params (extra->pbkdf, &loc_error);
+
+            if (pbkdf == NULL && loc_error != NULL) {
+                crypt_free (cd);
+                g_strfreev (cipher_specs);
+                bd_utils_report_finished (progress_id, loc_error->message);
+                g_clear_error (&loc_error);
```
Comment (tbzatek) [inline src/plugins/crypto.c]: Hmm, would it make sense to propagate `loc_error` to the parent `error` if we're returning `FALSE`?

**Pull request merged**